### PR TITLE
SDAP-506: STAC Catalog for Matchup outputs

### PR DESCRIPTION
**Pull request by skorper**
Ticket: [SDAP-506](https://issues.apache.org/jira/browse/SDAP-506)

Made the following changes:

* Added new `/cdmscatalog` endpoint
   * `/cdmscatalog/{id}` is a STAC catalog
   * `/cdmscatalog/{id}/{format}` is a STAC collection
     * In the `links` section of this output, there are links to all pages of `/cdmsresults` for that execution ID. 
     * `pageSize` is an optional param. Default is 1000 if not provided, If provided, it will be used to calculate the `/cdmsresults` links on `/cdmscatalog/{id}/{format}`
* Added link to `/cdmscatalog` endpoint in `/job` endpoint links
* Updated openapi spec

---

Example outputs:

```bash
job?id=a15bfac6-f9c5-4a64-b8ec-bedbf955d8c9
```

```json
{
	"status": "success",
	"message": "",
	"createdAt": "2023-12-07 23:45:10.877000",
	"updatedAt": "2023-12-07 23:49:23.176000",
	"links": [{
		"href": "https://localhost:8083/job?id=a15bfac6-f9c5-4a64-b8ec-bedbf955d8c9",
		"title": "The current page",
		"type": "application/json",
		"rel": "self"
	}, {
		"href": "https://localhost:8083/cdmscatalog/a15bfac6-f9c5-4a64-b8ec-bedbf955d8c9",
		"title": "STAC Catalog for execution results",
		"type": "application/json",
		"rel": "stac"
	}, {
		"href": "https://localhost:8083/cdmsresults?id=a15bfac6-f9c5-4a64-b8ec-bedbf955d8c9&output=CSV",
		"title": "Download results",
		"type": "text/csv",
		"rel": "data"
	}, {
		"href": "https://localhost:8083/cdmsresults?id=a15bfac6-f9c5-4a64-b8ec-bedbf955d8c9&output=JSON",
		"title": "Download results",
		"type": "application/json",
		"rel": "data"
	}, {
		"href": "https://localhost:8083/cdmsresults?id=a15bfac6-f9c5-4a64-b8ec-bedbf955d8c9&output=NETCDF",
		"title": "Download results",
		"type": "binary/octet-stream",
		"rel": "data"
	}],
	"params": {
		"primary": "MUR25-JPL-L4-GLOB-v04.2",
		"matchup": "AC2019-JPL",
		"startTime": "2019-01-14 00:00:00+00:00",
		"endTime": "2019-08-05 23:59:59+00:00",
		"bbox": "-180,-62,180,-44",
		"timeTolerance": 86400,
		"radiusTolerance": 25000,
		"platforms": "3B",
		"parameter": "sea_water_temperature",
		"depthMin": -20,
		"depthMax": 10
	},
	"jobID": "a15bfac6-f9c5-4a64-b8ec-bedbf955d8c9"
}
```

```bash
/cdmscatalog/a15bfac6-f9c5-4a64-b8ec-bedbf955d8c9
```

```json
{
	"stac_version": "1.0.0",
	"type": "Catalog",
	"id": "a15bfac6-f9c5-4a64-b8ec-bedbf955d8c9",
	"description": "STAC Catalog for CDMS output",
	"links": [{
		"rel": "collection",
		"href": "https://localhost:8083/cdmscatalog/a15bfac6-f9c5-4a64-b8ec-bedbf955d8c9/CSV",
		"title": "Collection of pages for a15bfac6-f9c5-4a64-b8ec-bedbf955d8c9 CSV output"
	}, {
		"rel": "collection",
		"href": "https://localhost:8083/cdmscatalog/a15bfac6-f9c5-4a64-b8ec-bedbf955d8c9/JSON",
		"title": "Collection of pages for a15bfac6-f9c5-4a64-b8ec-bedbf955d8c9 JSON output"
	}, {
		"rel": "collection",
		"href": "https://localhost:8083/cdmscatalog/a15bfac6-f9c5-4a64-b8ec-bedbf955d8c9/NETCDF",
		"title": "Collection of pages for a15bfac6-f9c5-4a64-b8ec-bedbf955d8c9 NETCDF output"
	}]
}
```
```bash
/cdmscatalog/a15bfac6-f9c5-4a64-b8ec-bedbf955d8c9/JSON
```

```json
{
	"stac_version": "1.0.0",
	"type": "Collection",
	"license": "not-provided",
	"id": "a15bfac6-f9c5-4a64-b8ec-bedbf955d8c9.JSON",
	"description": "Collection of results for CDMS execution and result format",
	"extent": {
		"spatial": {
			"bbox": [
				-180,
				-62,
				180,
				-44
			]
		},
		"temporal": {
			"interval": [
				"2019-01-14T00:00:00+00:00",
				"2019-08-05T23:59:59+00:00"
			]
		}
	},
	"links": [{
		"rel": "self",
		"href": "https://localhost:8083/cdmscatalog/a15bfac6-f9c5-4a64-b8ec-bedbf955d8c9/JSON",
		"title": "The current page",
		"type": "application/json"
	}, {
		"rel": "root",
		"href": "https://localhost:8083/cdmscatalog/a15bfac6-f9c5-4a64-b8ec-bedbf955d8c9",
		"title": "Root catalog for a15bfac6-f9c5-4a64-b8ec-bedbf955d8c9"
	}, {
		"rel": "data",
		"href": "https://localhost:8083/cdmsresults?id=a15bfac6-f9c5-4a64-b8ec-bedbf955d8c9&output=JSON&pageNum=1&pageSize=1000"
	}, {
		"rel": "data",
		"href": "https://localhost:8083/cdmsresults?id=a15bfac6-f9c5-4a64-b8ec-bedbf955d8c9&output=JSON&pageNum=1001&pageSize=1000"
	}, {
		"rel": "data",
		"href": "https://localhost:8083/cdmsresults?id=a15bfac6-f9c5-4a64-b8ec-bedbf955d8c9&output=JSON&pageNum=2001&pageSize=1000"
	}, {
		"rel": "data",
		"href": "https://localhost:8083/cdmsresults?id=a15bfac6-f9c5-4a64-b8ec-bedbf955d8c9&output=JSON&pageNum=3001&pageSize=1000"
	}, {
		"rel": "data",
		"href": "https://localhost:8083/cdmsresults?id=a15bfac6-f9c5-4a64-b8ec-bedbf955d8c9&output=JSON&pageNum=4001&pageSize=1000"
	}, {
		"rel": "data",
		"href": "https://localhost:8083/cdmsresults?id=a15bfac6-f9c5-4a64-b8ec-bedbf955d8c9&output=JSON&pageNum=5001&pageSize=1000"
	}]
}
```

Example with different page size:

```bash
/cdmscatalog/a15bfac6-f9c5-4a64-b8ec-bedbf955d8c9/JSON?pageSize=2000
```

```json
{
	"stac_version": "1.0.0",
	"type": "Collection",
	"license": "not-provided",
	"id": "a15bfac6-f9c5-4a64-b8ec-bedbf955d8c9.JSON",
	"description": "Collection of results for CDMS execution and result format",
	"extent": {
		"spatial": {
			"bbox": [
				-180,
				-62,
				180,
				-44
			]
		},
		"temporal": {
			"interval": [
				"2019-01-14T00:00:00+00:00",
				"2019-08-05T23:59:59+00:00"
			]
		}
	},
	"links": [{
		"rel": "self",
		"href": "https://localhost:8083/cdmscatalog/a15bfac6-f9c5-4a64-b8ec-bedbf955d8c9/JSON",
		"title": "The current page",
		"type": "application/json"
	}, {
		"rel": "root",
		"href": "https://localhost:8083/cdmscatalog/a15bfac6-f9c5-4a64-b8ec-bedbf955d8c9",
		"title": "Root catalog for a15bfac6-f9c5-4a64-b8ec-bedbf955d8c9"
	}, {
		"rel": "data",
		"href": "https://localhost:8083/cdmsresults?id=a15bfac6-f9c5-4a64-b8ec-bedbf955d8c9&output=JSON&pageNum=1&pageSize=2000"
	}, {
		"rel": "data",
		"href": "https://localhost:8083/cdmsresults?id=a15bfac6-f9c5-4a64-b8ec-bedbf955d8c9&output=JSON&pageNum=2001&pageSize=2000"
	}, {
		"rel": "data",
		"href": "https://localhost:8083/cdmsresults?id=a15bfac6-f9c5-4a64-b8ec-bedbf955d8c9&output=JSON&pageNum=4001&pageSize=2000"
	}]
}
```

### CODE REVIEW

File: analysis/webservice/algorithms/doms/ResultsStorage.py [hunk 1]
```diff
@@ -357,19 +357,18 @@ def __rowToDataEntry(self, row, trim_data=False):
 
         return entry
 
-    def __retrieveStats(self, id):
-        cql = "SELECT num_gridded_matched, num_insitu_matched, time_to_complete FROM doms_execution_stats where execution_id = %s limit 1"
+    def retrieveStats(self, id):
```
Comment (skorper) [inline analysis/webservice/algorithms/doms/ResultsStorage.py:360]: NOTE: This change was originally made in another PR that hasn't yet been merged. Will rebase once we merged PRs and this should not show up in the diff.

**Comment by skorper** [comment]
I just noticed the `pageNum` values in the stac collection outputs are incorrect. I'll need to fix that before we merge.

File: analysis/webservice/algorithms/doms/StacCatalog.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+                'title': f'Root catalog for {execution_id}',
+            }
+        ]
+
+        url = f'https://{self.host}/cdmsresults?id={execution_id}&output={output_format}'
+        for page_num in range(1, num_primary_matched, page_size):
+            links.append({
+                'rel': 'data',
+                'href': f'{url}&pageNum={page_num}&pageSize={page_size}'
+            })
```
Comment (RKuttruff) [inline analysis/webservice/algorithms/doms/StacCatalog.py:88]: If the matchup has only one primary match, this loop will miss the results link.
Ex: `0e664875-61fe-4540-a80d-1fba8f1fad61`

`/job`:
```json
{
    "status": "success",
    "message": "",
    "createdAt": "2024-02-01 19:19:37.474000",
    "updatedAt": "2024-02-01 19:19:46.084000",
    "links": [
        {
            "href": "https://doms.jpl.nasa.gov/job?id=0e664875-61fe-4540-a80d-1fba8f1fad61",
            "title": "Get job status - the current page",
            "type": "application/json",
            "rel": "self"
        },
        {
            "href": "https://doms.jpl.nasa.gov/cdmscatalog/0e664875-61fe-4540-a80d-1fba8f1fad61",
            "title": "STAC Catalog for execution results",
            "type": "application/json",
            "rel": "stac"
        },
        {
            "href": "https://doms.jpl.nasa.gov/cdmsresults?id=0e664875-61fe-4540-a80d-1fba8f1fad61&output=CSV",
            "title": "Download CSV results",
            "type": "text/csv",
            "rel": "data"
        },
        {
            "href": "https://doms.jpl.nasa.gov/cdmsresults?id=0e664875-61fe-4540-a80d-1fba8f1fad61&output=JSON",
            "title": "Download JSON results",
            "type": "application/json",
            "rel": "data"
        },
        {
            "href": "https://doms.jpl.nasa.gov/cdmsresults?id=0e664875-61fe-4540-a80d-1fba8f1fad61&output=NETCDF",
            "title": "Download NETCDF results",
            "type": "binary/octet-stream",
            "rel": "data"
        }
    ],
    "params": {
        "primary": "JPL-L4-MRVA-CHLA-GLOB-v3.0",
        "matchup": "shark-2018",
        "startTime": "2018-04-01 00:00:00+00:00",
        "endTime": "2018-04-01 23:59:59+00:00",
        "bbox": "-131,26,-130,27",
        "timeTolerance": 86400,
        "radiusTolerance": 10000.0,
        "platforms": "3B",
        "parameter": "mass_concentration_of_chlorophyll_in_sea_water",
        "depthMin": -5.0,
        "depthMax": 5.0
    },
    "executionID": "0e664875-61fe-4540-a80d-1fba8f1fad61",
    "totalPrimaryMatched": 1,
    "totalSecondaryMatched": 1,
    "averageSecondaryMatched": 1,
    "totalUniqueSecondaryMatched": 1
}
```

`/cdmscatalog/0e664875-61fe-4540-a80d-1fba8f1fad61/JSON`:
```json
{
    "stac_version": "1.0.0",
    "type": "Collection",
    "license": "not-provided",
    "id": "0e664875-61fe-4540-a80d-1fba8f1fad61.JSON",
    "description": "Collection of results for CDMS execution and result format",
    "extent": {
        "spatial": {
            "bbox": [
                -131.0,
                26.0,
                -130.0,
                27.0
            ]
        },
        "temporal": {
            "interval": [
                "2018-04-01T00:00:00+00:00",
                "2018-04-01T23:59:59+00:00"
            ]
        }
    },
    "links": [
        {
            "rel": "self",
            "href": "https://doms.jpl.nasa.gov/cdmscatalog/0e664875-61fe-4540-a80d-1fba8f1fad61/JSON",
            "title": "The current page",
            "type": "application/json"
        },
        {
            "rel": "root",
            "href": "https://doms.jpl.nasa.gov/cdmscatalog/0e664875-61fe-4540-a80d-1fba8f1fad61",
            "title": "Root catalog for 0e664875-61fe-4540-a80d-1fba8f1fad61"
        }
    ]
}
```

**Review by RKuttruff — DISMISSED** [review]
Found a minor bug when matchups have 1 result

**Review by ngachung — APPROVED** [review]
This change has been tested with a deployment to CDMS cluster and looks good.

**Pull request merged**